### ISSUE TITLE
StorageClass complexity for static Volumes (Ceph or Longhorn)

### ISSUE THREAD

**Issue by rsoika** [kubernetes/kubernetes#103819]
#### What would you like to be added:

I have a question/feature request about the usage of static volumes.

I am running a self managed kubernetes cluster (v1.21.2). For persistence I am running [Longhorn](https://longhorn.io/) inside my cluster and also a separate [ceph cluster ](https://ceph.io) running outside of my cluster and connecting via the Ceph [ceph-csi plugin](https://github.com/ceph/ceph-csi). Everything works fine.

To define a static volume I currently do the following

1) First I create in my ceph cluster or in lonhorn an image with a specific name and size. This is fine.

2) In the next step I have to create in my kubernetes Deployment a PersistenceVolume and PVC. 

For both - Ceph and Longhorn - the PersistenceVolume seems to be unnecessary complex. For example in ceph it has to look like this:

```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: my-dbdata
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 10Gi
  csi:
    driver: rbd.csi.ceph.com
    fsType: ext4
    nodeStageSecretRef:
      name: csi-rbd-secret
      namespace: ceph-system
    volumeAttributes:
      "clusterID": "1xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
      "pool": "kubernetes"
      "staticVolume": "true"
      "imageFeatures": "layering"
    volumeHandle: "my-dbdata"
  persistentVolumeReclaimPolicy: Retain
  volumeMode: Filesystem
```

I wonder why I (as an cluster admin) can't simply define a storage class with the csi cluster data
```

  csi:
    driver: rbd.csi.ceph.com
    fsType: ext4
    nodeStageSecretRef:
      name: csi-rbd-secret
      namespace: ceph-system
    volumeAttributes:
      "clusterID": "1xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
      "pool": "kubernetes"
      "staticVolume": "true"
      "imageFeatures": "layering"
  persistentVolumeReclaimPolicy: Retain
  volumeMode: Filesystem
```

...and than reduce my PersitenceVolume to a more simple config:

```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: my-dbdata
spec:
  storageClassName: ceph-durable
  volumeHandle: "my-dbdata"
```

Also in Longhorn a static persistence volume looks currently like this containing lot of csi details:

```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: my-dbdata
spec:
  capacity:
    storage: 100Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteOnce
  claimRef:
    namespace: default
    name: dbdata
  csi:
    driver: driver.longhorn.io
    fsType: ext4
    volumeHandle: my-db
  storageClassName: longhorn-durable
```

In longhorn you can define a storage class, but even here I need to provice CSI data like the driver name for each PersistenceVolume.

I asked the question first in the [ceph-csi community](https://github.com/ceph/ceph-csi/issues/2299) and the answer was that this is a restriction from Kubernetes and not form Ceph.

#### Why is this needed:

To sum it up: It would be easier for projects in my Kubernetes Cluster if a PersitenceVolume could be defined like this:

```
kind: PersistentVolume
apiVersion: v1
metadata:
  name: my-dbdata
spec:
  storageClassName: ceph-durable
  volumeHandle: "my-dbdata"
```
The project should not know the details about the CSI drivers and storage cluster IDs or capacity. For a static volume I should be sufficient to define the storageClassName and the volumeHandle to define the storage image uniquely.

What did you think about this?

**Comment by humblec** [comment]
>
The project should not know the details about the CSI drivers and storage cluster IDs or capacity. For a static volume I should be sufficient to define the storageClassName and the volumeHandle to define the storage image uniquely.

What did you think about this?
>

The `PV` object  and `SC` objects are considered `admin` objects and `PVC` is a `user` object . With that in mind, the details of the underlying storage is supposed to be abstracted from the user and in above terms, your `project`.   The storage class and its parameter parsing is done while the provisioning request is handled by a specific dynamic provisioner/driver.  In case of static binding the storage details are not parsed and handed over to the provisioner, , however for the PV controller to work ro do binding , the "spec" of a particular storage is required so the need comes in to fill it under `PV->SPEC->DRIVER(csi)->`. 

@jsafrane @msau42 @xing-yang any other thoughts to add ?

**Comment by rsoika** [comment]
just to give one more thought from my side. I am running a small self hosted kubernetes cluster. We are a small company. So I am the admin and the user of a project in one person. 

I run many projects on my cluster. And using ceph I need to setup the Ceph cluster details in every PV. I agree that the project user should not deal with the CSI details (even in my case it's the same person). And the PVC is not the problem, all is fine here. But defining a PV with the help of a storage class would make the PV easier as I mentioned in the example above. In a minimalistic configuration, the PV should only know the storageClass (holding the ceph csi details) and the Image name (specific to the pv)

**Comment by msau42** [comment]
Handling references to another object become complex when that object disappears or mutates.

It sounds like what you're interested in is a way to template a PV. Would something like kustomize or some other templating tool help with your use case?

**Comment by rsoika** [comment]
But isn't a storageClass exactly the object to handle the storage details for a PV? 
I wonder why the ceph-csi team moves the question in your direction. See the discussion [here](https://github.com/ceph/ceph-csi/issues/2299).

Or to ask in another way: Why should StorageClasses  not be used for static persistence volumes? The longhorn project does it exactly in this way.
